packages/plugin-hardhat/src/index.ts

```diff
@@ -0,0 +1,40 @@
+/* eslint-disable @typescript-eslint/no-var-requires */
+
+import '@nomiclabs/hardhat-ethers';
+import './type-extensions';
+import { subtask, extendEnvironment } from 'hardhat/config';
+import { TASK_COMPILE_SOLIDITY_COMPILE } from 'hardhat/builtin-tasks/task-names';
+import { lazyObject } from 'hardhat/plugins';
+import { validate, solcInputOutputDecoder, SolcInput } from '@openzeppelin/upgrades-core';
+import { writeValidations } from './validations';
+
+interface RunCompilerArgs {
+  input: SolcInput;
+}
+
+subtask(TASK_COMPILE_SOLIDITY_COMPILE, async (args: RunCompilerArgs, hre, runSuper) => {
```

Comment:
We have to check whether Harhdat can run multiple compilations concurrently. In this case there is a race condition with `writeValidations`. Not sure yet how to solve it if it is indeed a problem.

Reply:
Just confirmed with Hardhat team that compilations don't run concurrently now, but may in the future.